### PR TITLE
Further clarify stream class, and align better with C++ ostream in the process

### PR DESCRIPTION
**Pull request by mkinsner**
- rename constructor members to be more descriptive
- use flush instead of statement-based descriptions.  Flush is the C++ approach to describe/define
- endl implicitly flushes, like C++ convention
- no OpenCL interop on stream, since no appropriate object to map
- clarify synchronization and cross-WI guarantees
- improve wording

### CODE REVIEW

File: latex/programming_interface.tex [hunk 1]
```diff
@@ -3770,16 +3779,19 @@ \subsection{Stream class interface}
...
+An instance of the SYCL \codeinline{stream} class is required to synchronize with the host, and must output everything that is streamed to it via the \codeinline{operator<<()} operator (that doesn't exceed the \codeinline{workItemBufferSize} or \codeinline{totalBufferSize} limits) within a SYCL kernel function by the time that the event associated with a command group submission enters the completed state. The point at which this synchronization occurs and the method by which this synchronization is performed are implementation defined. For example it is valid for an implementation to use \codeinline{printf()}.
```
Comment (AerialMantis) [inline latex/programming_interface.tex]: When specifying what is required to be synchronised with the host, we should specify that it's everything that is streamed via `operator<<` before performing a flush. Clarifying that if you do not flush, there is no guarantee that anything is output.

File: latex/programming_interface.tex [hunk 2]
```diff
@@ -3559,13 +3559,13 @@ \section{Synchronization and atomics}
...
 
-All member functions of the \codeinline{stream} class are synchronous and errors are handled by throwing synchronous SYCL exceptions.
+If the \codeinline{totalBufferSize} or \codeinline{workItemBufferSize} limits are exceeded, it is implementation defined whether the streamed characters exceeding the limit are output, or silently ignored/discarded, and if output it is implementation defined whether those extra characters exceeding the \codeinline{workItemBufferSize} limit count toward the \codeinline{totalBufferSize} limit.  Regardless of this implementations defined behavior of output exceeding the limits, no undefined or erroneous behavior is permitted of an implementation when the limits are exceeded.  Unused characters within \codeinline{workItemBufferSize} (any portion of the \codeinline{workItemBufferSize} capacity that has not been used at the time of a stream flush) do not count toward the \codeinline{totalBufferSize} limit, in that only characters flushed count toward the \codeinline{totalBufferSize} limit.
```
[the remark this answers is not in the document]
Comment (mkinsner) [reply latex/programming_interface.tex]: Yes, correct.  We don't want to mandate that characters beyond the limit are output, but we also don't want to allow crashes or data corruption, because that won't be usable and this is mostly a debug feature.  Do we need to clarify this wording in some way?

File: latex/headers/stream.h [hunk 3]
```diff
@@ -47,12 +51,16 @@ __width_manipulator__ setw(int width);
...
  public:
 
-  stream(size_t bufferSize, size_t maxStatementSize, handler& cgh);
+  stream(size_t totalBufferSize, size_t workItemBufferSize, handler& cgh);
 
   /* -- common interface members -- */
 
   size_t get_size() const;
 
+  size_t get_work_item_buffer_size() const;
```
[the remark this answers is not in the document]
Comment (mkinsner) [reply latex/headers/stream.h]: No.  Intent is to rename `get_max_statement_size()` to `get_work_item_buffer_size()` which is aligned with the constructor terminology.  It still returns the value passed as the first constructor arg.

[the remark this answers is not in the document]
Comment (mkinsner) [reply latex/headers/stream.h]: Perhaps that counts as deprecating, but in a weaker way than removal of a feature.

File: latex/programming_interface.tex [hunk 4]
```diff
@@ -3636,11 +3636,20 @@ \subsection{Stream class interface}
...
+  stream buffer, and that empties the work item stream buffer.  After a flush, the full
+  \codeinline{workItemBufferSize} is available again for subsequent streaming within the work item.
+}
+\addRow
 {
   endl
 }
 {
-  Outputs a new-line character.
+  Outputs a new-line character and then generates a flush.
```
[the remark this answers is not in the document]
Comment (mkinsner) [reply latex/programming_interface.tex]: I've defined it near the beginning of the section, before the table, with the latest commit.

File: latex/programming_interface.tex [hunk 5]
```diff
@@ -3753,12 +3762,12 @@ \subsection{Stream class interface}
   \addRow
     {size_t get_size() const}
     {
-      Returns the maximum buffer size.
+      Returns the total buffer size, in characters.
     }
   \addRow
-    {size_t get_max_statement_size() const}
+    {size_t get_work_item_buffer_size() const}
```
[the remark this answers is not in the document]
Comment (mkinsner) [reply latex/programming_interface.tex]: Done

File: latex/programming_interface.tex (hunk 1 above)
Comment (mkinsner) [reply latex/programming_interface.tex]: I added the wording. Note that we've said that there is an implicit flush on kernel completion, defined in the "Implicit flush" section near the end of the stream section.  We believe that the implicit flush on kernel completion is important for usability.

File: latex/programming_interface.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (AerialMantis) [reply latex/programming_interface.tex]: Great, no I think this is clear, I just wanted to make sure I was interpreting that correctly, thanks.

File: latex/headers/stream.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (AerialMantis) [reply latex/headers/stream.h]: I think this makes sense, if by renaming it we at some point remove `get_max_statement_size` then I would consider this to be a deprecation, simply because it is available in SYCL 1.2.1 but would not be available in a later revision.

To clarify I am fine with this, but if we are deprecating an interface I think we should mark it as deprecated in the specification so our intentions are clear.

File: latex/programming_interface.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (AerialMantis) [reply latex/programming_interface.tex]: This looks good now thanks, though we can probably simplify the wording of the `flush` manipulator since it's defined earlier.

File: latex/programming_interface.tex (hunk 1 above)
Comment (AerialMantis) [reply latex/programming_interface.tex]: I agree this would be semantically preferable, though this requirement could infer the implementation strategy because in order for this to happen the per-work-item buffer would have to outlive the kernel function. Which either means copying back the collective buffer that contains each work item's buffer or inserting an implicit flush at the end of the kernel function.

File: latex/programming_interface.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkinsner) [reply latex/programming_interface.tex]: I did try simplifying the flush manipulator description, but then readded because I thought it was too hard to understand with the distributed wording.  I don't really mind either way - do you want me to remove the description from the table?

File: latex/programming_interface.tex (hunk 1 above)
Comment (mkinsner) [reply latex/programming_interface.tex]: Yes, understood.  Are you suggesting that there should be no implicit flush on kernel completion because it constrains implementations too much, or just noting that there are implementation ramifications?

File: latex/headers/stream.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkinsner) [reply latex/headers/stream.h]: I think I clarified appropriately with the latest commit.
